Kibana 6.4 forces nodejs 8.11.4 at runtime

kakra opened the pull request:
An alternative would be to not remove the bundled nodejs...

```
● kibana.service - Kibana Web Application
   Loaded: loaded (/usr/lib/systemd/system/kibana.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Sun 2018-09-16 07:01:21 CEST; 6h ago
  Process: 16613 ExecStart=/opt/kibana/bin/kibana --config /etc/kibana/kibana.yml --path.data=/var/lib/kibana (code=exited, status=1/FAILURE)
 Main PID: 16613 (code=exited, status=1/FAILURE)

Sep 16 07:01:20 kibana01 systemd[1]: Started Kibana Web Application.
Sep 16 07:01:21 kibana01 kibana[16613]: Kibana does not support the current Node.js version v8.11.1. Please use Node.js v8.11.4.
Sep 16 07:01:21 kibana01 systemd[1]: kibana.service: Main process exited, code=exited, status=1/FAILURE
Sep 16 07:01:21 kibana01 systemd[1]: kibana.service: Failed with result 'exit-code'.
```

kakra commented:
I'm not sure why this CI check shows up, kibana-bin depends on keyworded nodejs, yes... But it's keyworded itself. I needed to use `repoman ci -i` to ignore this problem.

Still learning... Maybe someone sheds light on this? :-)

hydrapolic commented:
Thanks for the report, I had 8.11.4 so I didn't catch this locally. I think it will be better to leave nodejs unbundled and specify the version in RDEPEND (and stripping the nodejs check). It seems like 8.12.0 works fine with kibana 6.4.0.

hydrapolic commented:
I'll open a PR with the change.

hydrapolic commented:
The fix is available here: https://github.com/gentoo/gentoo/pull/9892

hydrapolic commented:
> I'm not sure why this CI check shows up, kibana-bin depends on keyworded nodejs, yes... But it's keyworded itself. I needed to use `repoman ci -i` to ignore this problem.

If you mean the failing CI check https://qa-reports.gentoo.org/output/gentoo-ci/61b9f15/output.html#www-apps/kibana-bin then it means that nodejs-8 is not available on hardened in 13.0 profile, since icu it requires is masked there :)

hydrapolic commented:
It works after adding a mask to 13.0 profile: https://github.com/gentoo/gentoo/pull/9892/commits/5c5d951d8d2b083953bd5cab249c802884fe2f41

kakra commented:
@hydrapolic Thanks but this report already was/is a PR... ;-)

kakra commented:
@hydrapolic But I see you're using a different approach. My PR adds the required nodejs version at least. Maybe it should be in portage in case someone really needs it for a plugin to work?

hydrapolic commented:
@kakra, you can of course convince the nodejs maintainer to revert the version drop if you have problems with nodejs 8.12.0. I've tested the OSS/x-pack version (just a basic check) and it seems ok.

hydrapolic commented:
> @hydrapolic I think that's a chicken/egg problem: It probably was dropped because nothing explicitly depended on it. If your tests are okay, I'm fine with 8.12.

If you want to bring nodejs 8.11.4 back then just revert https://github.com/gentoo/gentoo/commit/70fe9cb89f36a8075ed5ffdb9d6e5d10c10d0c86, it's not necessary to re-add the package again. You also need to bump kibana to new revision as it will be broken for existing users.

kakra commented:
> > @hydrapolic I think that's a chicken/egg problem: It probably was dropped because nothing explicitly depended on it. If your tests are okay, I'm fine with 8.12.
> 
> If you want to bring nodejs 8.11.4 back then just revert [70fe9cb](https://github.com/gentoo/gentoo/commit/70fe9cb89f36a8075ed5ffdb9d6e5d10c10d0c86), it's not necessary to re-add the package again. You also need to bump kibana to new revision as it will be broken for existing users.

Fixed and also re-ordered the commits to have a consistent portage at every step.

kakra commented:
@hydrapolic I've added a Co-authored-by to credit the ideas taken from your PR. I'm guessing that's okay.

I've now implemented an idea to opt-out of strict dependencies via a use-flag.

hydrapolic commented:
To be honest, we do this all the time in Gentoo. They have to bundle some version of node.js, but it doesn't mean other minor versions won't work.

Now it screams on different version because they don't expect you to unbundle node.js. But this is an artificial limitation because the time they released it, node 8.12 didn't exist. If we know a newer minor versiin of node 8 breaks kibana, let's add a hard constrain. But I think this is not the case (I'm running wiyh node 8.12).

hydrapolic commented:
The DEPEND/RDEPEND constrains should reflect reality where you must use a specific version, in this case it seems like it's unnecessary. If others report otherwise, let's do it, but it doesn't seem like the case :(

I'm not aware of using a "strict-deps" USE flag in Gentoo :-/

kakra commented:
> To be honest, we do this all the time in Gentoo. They have to bundle some version of node.js, but it doesn't mean other minor versions won't work.
> 
> Now it screams on different version because they don't expect you to unbundle node.js. But this is an artificial limitation because the time they released it, node 8.12 didn't exist. If we know a newer minor version of node 8 breaks kibana, let's add a hard constrain. But I think this is not the case (I'm running kibana 6.4 with node 8.12).

Okay, so it's more or less the old `dosver` problem we had in DOS some decades ago... I wonder why they place the constrain when they bundle nodejs anyways. Well, whatever it is, let's go with your solution then.

hydrapolic commented:
Haven't checked, but maybe they take the node.js version to bundle from that string. It would then make sense to have it one place.

hydrapolic commented:
> > To be honest, we do this all the time in Gentoo. They have to bundle some version of node.js, but it doesn't mean other minor versions won't work.
> > Now it screams on different version because they don't expect you to unbundle node.js. But this is an artificial limitation because the time they released it, node 8.12 didn't exist. If we know a newer minor version of node 8 breaks kibana, let's add a hard constrain. But I think this is not the case (I'm running kibana 6.4 with node 8.12).
> 
> Okay, so it's more or less the old `dosver` problem we had in DOS some decades ago... I wonder why they place the constrain when they bundle nodejs anyways. Well, whatever it is, let's go with your solution then.

You can close this PR in that case :)
